include/net/mqtt_sn.h

```diff
@@ -0,0 +1,197 @@
+ * Targets protocol version 1.2.
+ *
+ */
+
+#ifndef ZEPHYR_INCLUDE_NET_MQTT_SN_H_
+#define ZEPHYR_INCLUDE_NET_MQTT_SN_H_
+
+#include <stddef.h>
+
+#include <zephyr.h>
+#include <zephyr/types.h>
+#include <net/buf.h>
+#include <net/net_ip.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * Quality of Service. QoS 0-2 work the same as basic MQTT, QoS -1 is an MQTT-SN addition.
+ */
+enum __attribute__((packed)) mqtt_sn_qos {
+	MQTT_SN_QOS_0, /**< QOS 0 */
+	MQTT_SN_QOS_1, /**< QOS 1 */
+	MQTT_SN_QOS_2, /**< QOS 2 */
+	MQTT_SN_QOS_M1 /**< QOS -1 */
+};
+
+enum __attribute__((packed)) mqtt_sn_topic_type {
+	MQTT_SN_TOPIC_TYPE_NORMAL,
+	MQTT_SN_TOPIC_TYPE_PREDEF,
+	MQTT_SN_TOPIC_TYPE_SHORT
+};
+
+enum __attribute__((packed)) mqtt_sn_return_code {
+	MQTT_SN_CODE_ACCEPTED = 0x00, /**< Accepted */
+	MQTT_SN_CODE_REJECTED_CONGESTION = 0x01, /**< Rejected: congestion */
+	MQTT_SN_CODE_REJECTED_TOPIC_ID = 0x02, /**< Rejected: Invalid Topic ID */
+	MQTT_SN_CODE_REJECTED_NOTSUP = 0x03, /**< Rejected: Not Supported */
+};
+
+/** @brief Abstracts memory buffers. */
+struct mqtt_sn_data {
+	const uint8_t *data; /**< Pointer to data. */
+	uint16_t size; /**< Size of data, in bytes. */
+};
+
+/**
+ * @brief Initialize memory buffer from C literal string.
+ *
+ * Use it as follows:
+ *
+ * struct mqtt_sn_data password = MQTT_SN_DATA_STRING_LITERAL("my_pass");
+ *
+ * @param[in] literal Literal string from which to generate mqtt_sn_data object.
+ */
+#define MQTT_SN_DATA_STRING_LITERAL(literal) ((struct mqtt_sn_data){ literal, sizeof(literal) - 1 })
+
+/**
+ * @brief Initialize memory buffer from single bytes.
+ *
+ * Use it as follows:
+ *
+ * struct mqtt_sn_data password = MQTT_SN_DATA_BYTES(0x13, 0x37);
+ */
+#define MQTT_SN_DATA_BYTES(...)                                                                    \
+	((struct mqtt_sn_data){ (uint8_t[]){ __VA_ARGS__ }, sizeof((uint8_t[]){ __VA_ARGS__ }) })
+
+/**
+ * Forward declaration. Library will return pointers to this type only.
+ */
+struct mqtt_sn_msg;
+
+/**
+ * @brief Increase the ref-count of an MQTT-SN message.
+ * 
+ * Use this function in the msg_send callback when you need the associated memory
+ * to remain valid until after the callback returns.
+ * Don't forget to also call mqtt_sn_msg_unref when making use of that functionality.
+ */
+struct mqtt_sn_msg *mqtt_sn_msg_ref(struct mqtt_sn_msg *msg);
+
+/**
+ * @brief Decrease the ref-count of an MQTT-SN message.
+ * 
+ * When the ref-count drops to 0, the message will be de-allocated.
+ */
+void mqtt_sn_msg_unref(struct mqtt_sn_msg *msg);
+
+/**
+ * Get a pointer to the message's content.
+ */
+void mqtt_sn_msg_data(struct mqtt_sn_msg *msg, struct mqtt_sn_data *data);
+
+/**
+ * @brief Callback structure to be used with mqtt_sn_client_init.
+ */
+struct mqtt_sn_callbacks {
+	/** Will be called by the library when it wants to send a message. */
+	int (*msg_send)(struct mqtt_sn_msg *msg, const void *recipient);
```

Comment:
I see that the approach here regarding communication (socket) handling is slightly different than in the core MQTT lib we have. Here, we expect the application to handle the communication part, while the standard MQTT lib handles sockets on its own, with an abstracted transport layer. I wonder if it made sense to follow that approach here, to give users unanimous feel of both libraries.

Reply:
This was a design decision I made because MQTT-SN does not require any specific kind of transport, unlike MQTT which does require a reliable transport like TCP / TLS. You can use UDP for MQTT-SN, but you could also use ZigBee, UART, GATT, LoRaWAN or potentially even SMS.

It might make sense to take inspiration from the original MQTT library to have some kind of transport-related struct of function pointers, and provide a default one for UDP. That would make the approach similar, even if the details will differ.

Reply:
@rlubos The more I go into the details _how_ the MQTT library handles the transport, the less I like it, actually. The user is expected to call `mqtt_input` when there is new data. To know if that is the case, the user of course has to poll the socket. This breaks the separation of concerns IMO (see `zephyr/samples/net/cloud/google_iot_mqtt/src/protocol.c:215`, `wait_for_input`).

Reply:
Well, there was an argument that it should be possible to run MQTT along with other protocols on a single thread, therefore `mqtt_input()` is non-blocking and relies on the application to actually monitor the activity on a socket (so that it can add whatever sockets it wants to the `fds`). But perhaps it wouldn't be a bad idea to have some kind of `mqtt_wait()` function.